### PR TITLE
 prevent adding empty/duplicated regex from UI

### PR DESCRIPTION
**Pull request by khalil2535**
* Solve #5899
* Remove duplicated messages from Messages.properties

### CODE REVIEW

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s): 
- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff
- https://wiki.eclipse.org/EGit/User_Guide/Commit#Signing_off_commits

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java [hunk 1]
```diff
@@ -212,8 +226,13 @@ protected void init() {
         @Override
         protected boolean validateFields() {
             try {
-                Pattern.compile(getRegexTextField().getText(), Pattern.CASE_INSENSITIVE);
-            } catch (PatternSyntaxException e) {
+                String trimmedRegex = getRegexTextField().getText().trim();
+                if (trimmedRegex.trim().isEmpty()) return false; // to prevent accepting empty regex
```
Comment (kingthorin) [inline zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: This is syntactically correct but it would be best to use curly braces for consistency

Comment (kingthorin) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: Also why trim the trimmed value, is there a reason to apply it twice?

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 2]
```diff
@@ -223,9 +223,6 @@ api.options.addr.dialog.remove.text = Are you sure you want to remove the select
 api.options.addr.dialog.remove.button.cancel = Cancel
 api.options.addr.dialog.remove.button.confirm = Remove
 api.options.addr.dialog.remove.checkbox.label = Do not show this message again
-api.options.addr.table.header.enabled	= Enabled
-api.options.addr.table.header.regex	= Regex
-api.options.addr.table.header.value	= Address
```
Comment (kingthorin) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Why were these API keys/values removed?

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 3]
```diff
@@ -1106,8 +1103,8 @@ cfu.update.message.resourcesNotSavedAndActiveActions = <html>There are unsaved r
 cfu.update.message.resourcesNotSaved   = <html>The following resources have not been saved:<ul>{0}</ul>The changes might be lost if the update fails. Continue with update?</html>
 cfu.warn.addon.with.extensions.with.missing.requirements = Optional Add-ons:
 cfu.warn.addon.with.missing.requirements = Missing Requirements:
-cfu.warn.addon.with.missing.requirements.addon = A cyclic dependency was detected.
 cfu.warn.addon.with.missing.requirements.addon = Add-on "{0}"
+cfu.warn.addon.with.missing.requirements.addon.cyclic = A cyclic dependency was detected.
```
Comment (kingthorin) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Why was this changed with no corresponding usage change?

**Comment by khalil2535** [comment]
Done .. (with 2 force pushes 🤦‍♂️)

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 4]
```diff
@@ -2115,6 +2112,7 @@ multiple.options.regexes.dialog.remove.regex.button.confirm = Remove
 multiple.options.regexes.dialog.remove.regex.checkbox.label = Do not show this message again
 multiple.options.regexes.dialog.regex.invalid.title = Invalid Regular Expression
 multiple.options.regexes.dialog.regex.invalid.text = The provided regular expression is not valid:\n{0}
+multiple.options.regexes.dialog.regex.invalid.alreadyExist.text = regular expression already exists
```
Comment (kingthorin) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Please turn this into a proper sentence (caps and punctuation).

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 2 above)
Comment (khalil2535) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Those keys/values are duplicated..
[line 219](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties#L219) == [line 226](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties#L226)
[line 220](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties#L220) == [line 227](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties#L227)
[line 221](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties#L221) == [line 228](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties#L228)

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 3 above)
Comment (khalil2535) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: The key is used just once and i see the fisrt key/value is the correct in the case.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 2 above)
Comment (thc202) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: I'd suggest doing these changes in other PR for clarity.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 3 above)
Comment (thc202) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: This message is still needed though, the code needs to be changed as well.

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java (hunk 1 above)
Comment (khalil2535) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: to make sure that it's trimmed 😂
ok i'll fix that

Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: Better trim in `checkAndEnableConfirmButton()`, not worth enable the button if we are not going to accept it.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 3 above)
Comment (khalil2535) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: I searched and only place "cfu.warn.addon.with.missing.requirements.addon" is used is here, do you want me to change it [here](https://github.com/zaproxy/zaproxy/blob/cfb89797be118cc9231e81db0c7dc090aec5eafe/zap/src/main/java/org/zaproxy/zap/control/AddOnRunIssuesUtils.java#L232) ?

image

Comment (kingthorin) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Plus cfu has nothing to do with this PR

Comment (khalil2535) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Ok

Comment (thc202) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: In the `CYCLIC` case.

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java (hunk 1 above)
Comment (khalil2535) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: @thc202 did this good suggestion

**Comment by kingthorin** [comment]
I’m trying to review this on a cell phone remotely it looks good to me but could you add a screenshot of the warning dialog when a user tries to add a duplicate entry? (I’m curious if the getLocalizedMessage call returns something useful from the exception we’ve newly thrown. [Hopefully it falls back to the getMessage text if a localized message isn’t found])

**Comment by khalil2535** [comment]
image

**Comment by kingthorin** [comment]
Thanks!

File: zap/src/main/java/org/zaproxy/zap/model/Context.java [hunk 5]
```diff
@@ -343,17 +343,18 @@ private static void validateRegexs(List<String> regexs) {
      * Validates that the given regular expression is not {@code null} nor invalid.
      *
      * @param regex the regular expression to be validated
-     * @throws IllegalArgumentException if the regular expression is {@code null}.
+     * @throws IllegalArgumentException if the regular expression is {@code null} or Empty.
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/model/Context.java]: - The description of the method could be tweaked as well. `validateRegexs` should be updated accordingly.
- The empty should be lower case (same in the exception message).

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java [hunk 6]
```diff
@@ -212,8 +226,12 @@ protected void init() {
         @Override
         protected boolean validateFields() {
             try {
-                Pattern.compile(getRegexTextField().getText(), Pattern.CASE_INSENSITIVE);
-            } catch (PatternSyntaxException e) {
+                String trimmedRegex = getRegexTextField().getText().trim();
+                if (regexes != null && regexes.contains(trimmedRegex)) {
+                    throw new IllegalArgumentException(TEXT_ALREADY_EXIST_INVALID_REGEX_DIALOG);
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: - We should not use exceptions for normal/common control flow (when we can avoid it, they are expensive to create). This should show the dialogue and return.
 -  I'd suggest using other title (e.g. `Duplicated Regular Expression`) for the error and show without `The provided regular expression is not valid:`, seems unnecessary.

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java [hunk 7]
```diff
@@ -212,8 +226,12 @@ protected void init() {
         @Override
         protected boolean validateFields() {
             try {
-                Pattern.compile(getRegexTextField().getText(), Pattern.CASE_INSENSITIVE);
-            } catch (PatternSyntaxException e) {
+                String trimmedRegex = getRegexTextField().getText().trim();
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: For consistency in what we accept/show we should also trim in `performAction`.

**Comment by thc202** [comment]
This still does not prevent adding duplicates to the `Context`, one can still add them programmatically or through the ZAP API. I'm OK if this is just meant to address the UI but the commit message/PR should say that.

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java [hunk 8]
```diff
@@ -211,23 +227,23 @@ protected void init() {
...
                         this,
-                        MessageFormat.format(TEXT_INVALID_REGEX_DIALOG, e.getLocalizedMessage()),
-                        TITLE_INVALID_REGEX_DIALOG,
+                        TEXT_ALREADY_EXIST_INVALID_REGEX_DIALOG,
+                        TITLE_ALREADY_EXIST_INVALID_REGEX_DIALOG,
                         JOptionPane.INFORMATION_MESSAGE);
                 getRegexTextField().requestFocusInWindow();
                 return false;
             }
+            Pattern.compile(trimmedRegex, Pattern.CASE_INSENSITIVE);
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: This lost the catch and error message, was that on purpose?

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java [hunk 9]
```diff
@@ -89,7 +87,7 @@ public void setRegexes(List<String> regexes) {
     @Override
     public String showAddDialogue() {
         if (addDialog == null) {
-            addDialog = new DialogAddRegex(owner);
+            addDialog = new DialogAddRegex(owner, this.getRegexes());
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: It will have to set the list each time (same for modify), it might change after constructing the dialogue.

**Comment by thc202** [comment]
The modify dialogue needs to override `validateFields` to only do the base validations if the regex is different (to avoid showing the error that's duplicated).

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java (hunk 9 above)
Comment (khalil2535) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java:90]: do i have to create new object each time or to create `setRegexes` method in the `DialogAddRegex` class ?

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java (hunk 8 above)
Comment (khalil2535) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java:240]: No.

**Comment by khalil2535** [comment]
> The modify dialogue needs to override `validateFields` to only do the base validations if the regex is different (to avoid showing the error that's duplicated).

It only shows the error when the regex is duplicated and the button "modify" is pressed.

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java (hunk 9 above)
Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java:90]: I'd go with the setter.

**Comment by thc202** [comment]
> It only shows the error when the regex is duplicated and the button "modify" is pressed.

Exactly, but (IMO) it shouldn't, there's no duplication it's the regex itself. In that case ZAP should just accept it.

File: zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java (hunk 9 above)
Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java:92]: The new constructors could be reverted, they are no longer needed.

Comment (khalil2535) [reply zap/src/main/java/org/zaproxy/zap/view/MultipleRegexesOptionsPanel.java]: Done @thc202.
Good advice's, I hope doing more development with you in this project.

**Comment by thc202** [comment]
Thanks!

**Comment by thc202** [comment]
@khalil2535 how would you like to be credited (e.g. name, handle)?
https://www.zaproxy.org/docs/desktop/credits/#zap-extended-team

**Comment by khalil2535** [comment]
> @khalil2535 how would you like to be credited (e.g. name, handle)?
> https://www.zaproxy.org/docs/desktop/credits/#zap-extended-team

how didn't i see this comment, omg of course.

name : Mahmoud Khalil
handle(if i understand it well): @khalil2535
